source/common/jwt_authn/extractor.cc

```diff
@@ -0,0 +1,148 @@
+#include "common/jwt_authn/extractor.h"
+
+#include "common/common/utility.h"
+#include "common/http/utility.h"
+
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtAuthentication;
+
+namespace Envoy {
+namespace JwtAuthn {
+namespace {
+
+// The authorization header name.
+const std::string AuthorizationHeader = "authorization";
+
+// The authorization bearer prefix.
+const std::string BearerPrefix = "Bearer ";
+
+// The query parameter name to get JWT token.
+const std::string ParamAccessToken = "access_token";
+
+// A base JwtLocation object to store token and specified_issuers.
+class JwtLocationBase : public JwtLocation {
+public:
+  JwtLocationBase(const std::string& token, const std::set<std::string>& issuers)
+      : token_(token), specified_issuers_(issuers) {}
+
+  // Get the token string
+  const std::string& token() const override { return token_; }
+
+  // Check if an issuer has specified the location.
+  bool isIssuerSpecified(const std::string& issuer) const override {
+    return specified_issuers_.find(issuer) != specified_issuers_.end();
+  }
+
+private:
+  // Extracted token.
+  std::string token_;
+  // Stored issuers specified the location.
+  const std::set<std::string>& specified_issuers_;
+};
+
+// The JwtLocation for header extraction.
+class JwtHeaderLocation : public JwtLocationBase {
+public:
+  JwtHeaderLocation(const std::string& token, const std::set<std::string>& issuers,
+                    const Http::LowerCaseString& header)
+      : JwtLocationBase(token, issuers), header_(header) {}
+
+  void removeJwt(Http::HeaderMap& headers) const override { headers.remove(header_); }
+
+private:
+  // the header name the JWT is extracted from.
+  const Http::LowerCaseString& header_;
+};
+
+// The JwtLocation for param extraction.
+class JwtParamLocation : public JwtLocationBase {
+public:
+  JwtParamLocation(const std::string& token, const std::set<std::string>& issuers,
+                   const std::string& param)
+      : JwtLocationBase(token, issuers), param_(param) {}
+
+  void removeJwt(Http::HeaderMap&) const override {
+    // TODO(qiwzhang): remove JWT from parameter.
+  }
+
+private:
+  // the parameter key the JWT is extracted from.
+  const std::string& param_;
+};
+
+} // namespace
+
+Extractor::Extractor(const JwtAuthentication& config) {
+  for (const auto& rule : config.rules()) {
+    bool use_default = true;
```

Comment:
`bool use_default = rule.from_headers_size > 0 || rule.from_param_size() > 0`
then you don't need two if below.

Reply:
Done